Source/JavaScriptCore/dfg/DFGClobberSet.h

```diff
@@ -74,6 +74,30 @@ class ClobberSet {
     UncheckedKeyHashMap<AbstractHeap, bool> m_clobbers;
 };
 
+class PreciseClobberSet {
+public:
+    bool overlaps(const HeapLocation& location)
+    {
+        if (location.base() && m_clobberedBases.contains(location.base()))
+            return true;
+        if (location.index() && location.index().isNode() && location.index()->isInt32Constant())
+            return m_clobberedLocations.contains(location);
+        return false;
+    }
+
+    void add(const HeapLocation& location)
+    {
+        if (location.index()->isInt32Constant())
+            m_clobberedLocations.add(location);
+        else
+            m_clobberedBases.add(location.base());
+    }
+
+private:
+    HashSet<HeapLocation> m_clobberedLocations;
+    HashSet<Node*> m_clobberedBases;
```

Comment:
This is not precisely trackable as it is possible that different Node* is pointing at the same base (aliasing).